consensus/istanbul/proxy/enode_certificate.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 The Celo Authors
+// This file is part of the celo library.
+//
+// The celo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The celo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the celo library. If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"encoding/hex"
+
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// handleEnodeCertificateMsg is invoked by the proxy
+// It will verify that the message is from a validator that is within the
+// validator connection set and then forward it to the proxied validator.
+func (p *proxyEngine) handleEnodeCertificateMsg(peer consensus.Peer, payload []byte) (bool, error) {
```

Comment:
Could this function be combined somewhat with the function below? Mostly the logic in this function a bit confusing, especially given that the function below exists and the caller checks before calling whether the message came from the proxied peer.

Reply:
I modified and cleaned up this function a bit and made it clear that it's the enode cert handler from remote validators.

There doesn't seem to be much overlap between this function and the one below.  And I do like the fact that the functions are now very explicit in their purpose, so would prefer to keep it separate.